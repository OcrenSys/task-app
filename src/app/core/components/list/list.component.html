<mat-card class="card mt-2 mb-2">
  <mat-card-header class="card-header">
    <mat-card-title-group>
      <mat-card-title>Completadas</mat-card-title>
      <mat-card-subtitle
        >Lista de tareas completadas &nbsp; {{ count }}/{{
          total
        }}</mat-card-subtitle
      >
    </mat-card-title-group>
  </mat-card-header>

  <mat-card-content>
    <div class="table-content">
      <mat-list role="list">
        <mat-list-item
          role="listitem"
          *ngFor="let task of tasks$ | async; trackBy: taskTrackBy"
        >
          <p mat-line strikethrough>
            {{ task.title }}
          </p>
          <mat-divider [inset]="true"></mat-divider>
        </mat-list-item>
      </mat-list>
    </div>
  </mat-card-content>
</mat-card>
