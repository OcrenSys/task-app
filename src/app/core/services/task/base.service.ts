import {
  HttpClient,
  HttpHeaders,
} from '@angular/common/http';
import { Observable } from 'rxjs';
import { IBaseService } from '../../../shared/classes/Base.class';

export class BaseService<T> extends IBaseService<T> {
  constructor(protected readonly _httpClient: HttpClient) {
    super();
  }

  protected _create<T>(
    url: string,
    model: T,
  ): Observable<unknown> {
    const headers = new HttpHeaders().set(
      'Content-Type',
      'application/json; charset=utf-8'
    );

    return this._httpClient.post<T>(
      url,
      JSON.stringify(model),
      {
        headers,
      } || {}
    );
  }

  protected _update<T>(
    url: string,
    id: string,
    model: T,
  ): Observable<unknown> {
    const headers = new HttpHeaders().set(
      'Content-Type',
      'application/json; charset=utf-8'
    );

    return this._httpClient.put(
      `${url}/${id}`,
      JSON.stringify(model),
      {
        headers,
      } || {}
    );
  }

  protected _list<T>(url: string): Observable<unknown> {
    return this._httpClient.get(url);
  }

  protected _delete<T>(url: string): Observable<unknown> {
    return this._httpClient.delete(url);
  }
}
