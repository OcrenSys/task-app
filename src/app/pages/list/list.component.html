<ng-container>
  <mat-card class="card-header">
    <mat-card-header>
      <mat-card-title>Tareas</mat-card-title>
      <mat-card-subtitle>Lista</mat-card-subtitle>
    </mat-card-header>
    <mat-card-actions align="end">
      <button mat-raised-button color="primary" (click)="openDialog()">
        <mat-icon
          aria-hidden="false"
          aria-label="Example home icon"
          fontIcon="add"
        ></mat-icon>
        Nueva Tarea
      </button>
    </mat-card-actions>
  </mat-card>

  <app-task-list></app-task-list>
</ng-container>
