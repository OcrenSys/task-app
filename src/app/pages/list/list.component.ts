import { Component } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MatCardModule } from '@angular/material/card';
import { MatDialog, MatDialogModule } from '@angular/material/dialog';
import { MatIconModule } from '@angular/material/icon';
import { NavbarComponent } from 'src/app/core/components/navbar/navbar.component';
import { TaskFormComponent } from 'src/app/core/components/task-form/task-form.component';
import { TaskListComponent } from 'src/app/core/components/task-list/task-list.component';

@Component({
  selector: 'app-list',
  templateUrl: './list.component.html',
  styleUrls: ['./list.component.css'],
  standalone: true,
  imports: [
    MatButtonModule,
    MatIconModule,
    NavbarComponent,
    TaskListComponent,
    TaskFormComponent,
    MatDialogModule,
    MatCardModule,
  ],
})
export class ListComponent {
  constructor(public dialog: MatDialog) {}

  protected openCreateModal() {
    const dialogRef = this.dialog.open(TaskFormComponent, {
      data: {
        title: '',
        description: ' ',
      },
      panelClass: 'dialog-responsive',
    });

    dialogRef.afterClosed().subscribe((result) => {
      console.log('The dialog was closed');
      if (result) console.log('Result', result);
    });
  }
}
