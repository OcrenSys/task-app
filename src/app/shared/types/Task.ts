export type Task = {
  title: string;
  description: string;
  createdAt: string;
  updatedAt: string;
  status: 'activa' | 'pendiente' | 'completado';
};
