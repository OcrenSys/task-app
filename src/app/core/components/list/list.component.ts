import { Component, OnInit } from '@angular/core';
import { MatTableModule } from '@angular/material/table';
import { MatIconModule } from '@angular/material/icon';
import { MatButtonModule } from '@angular/material/button';
import { MatCheckboxModule } from '@angular/material/checkbox';
import { HttpClientModule } from '@angular/common/http';
import { MatCardModule } from '@angular/material/card';
import { AsyncPipe, NgFor } from '@angular/common';
import { MatListModule } from '@angular/material/list';
import { Observable, map, tap } from 'rxjs';
import { MatDividerModule } from '@angular/material/divider';
import { MatSnackBarModule } from '@angular/material/snack-bar';
import { StrikethroughDirective } from '../../directives/strikethrough/strikethrough.directive';
import { TruncatePipe } from '../../../shared/pipes/truncate/truncate.pipe';
import { TaskStore } from '../../store/task.store';
import { Task } from '../../../shared/classes/Task.class';

@Component({
  selector: 'task-list',
  standalone: true,
  imports: [
    HttpClientModule,
    MatTableModule,
    MatButtonModule,
    MatIconModule,
    MatCheckboxModule,
    TruncatePipe,
    MatCardModule,
    MatSnackBarModule,
    MatListModule,
    MatDividerModule,
    NgFor,
    AsyncPipe,
    StrikethroughDirective,
  ],
  templateUrl: './list.component.html',
  styleUrls: ['./list.component.css'],
})
export class ListComponent implements OnInit {
  protected tasks$: Observable<Task[]> = new Observable<Task[]>();
  protected count: number = 0;
  protected total: number = 0;

  constructor(private taskStore: TaskStore<Task>) {}

  ngOnInit(): void {
    this.tasks$ = this.taskStore.store().pipe(
      tap((task) => (this.total = task.length)),
      map((tasks) => tasks.filter((task: Task) => task.isCompleted)),
      tap((task) => (this.count = task.length))
    );
  }

  taskTrackBy(index: number, task: Task) {
    return task.id;
  }
}
