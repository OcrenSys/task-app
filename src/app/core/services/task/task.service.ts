import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { BaseService } from './base.service';
import {
  SERVER_DOMAIN,
  TASK_MODEL,
} from 'src/app/shared/constants/urls.constants';
import { Task } from 'src/app/shared/classes/Task.class';

@Injectable({
  providedIn: 'root',
})
export class TaskService<T extends Task> extends BaseService<T> {
  constructor(protected override readonly _httpClient: HttpClient) {
    super(_httpClient);
  }

  public create<T>(model: T): Observable<unknown> {
    return this._create<T>(`${SERVER_DOMAIN}/${TASK_MODEL}`, model);
  }

  public update<T>(id: string, model: T): Observable<unknown> {
    return this._update(`${SERVER_DOMAIN}/${TASK_MODEL}`, id, model);
  }

  public list<T>() {
    return this._list<T>(`${SERVER_DOMAIN}/${TASK_MODEL}`);
  }

  public delete<T>(id: string) {
    return this._delete<T>(`${SERVER_DOMAIN}/${TASK_MODEL}/${id}`);
  }
}
