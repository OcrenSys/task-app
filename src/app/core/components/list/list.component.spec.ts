import { ComponentFixture, TestBed } from '@angular/core/testing';

import { ListComponent } from './list.component';
import { TaskService } from '../../services/task/task.service';
import { Task } from '../../../shared/classes/Task.class';
import { Observable, map, of, tap } from 'rxjs';

describe('ListComponent', () => {
  let component: ListComponent;
  let fixture: ComponentFixture<ListComponent>;
  const tasksMock: Task[] = [
    {
      id: 'LpWnRtlbUdc33o3Ih5ZY',
      createdAt: '1694674453597',
      description:
        'Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going through the cites of the word in classical literature, discovered the undoubtable source.',
      title: 'Where does it come from?',
      isCompleted: false,
      updatedAt: '1694725216717',
    },
    {
      id: 'PJEMW8RYJHsFwBtZZi6z',
      createdAt: '1694674465536',
      description:
        'It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout.',
      title: 'Where can I get some?',
      isCompleted: true,
      updatedAt: '1694755029652',
    },
    {
      id: 'PQIc81GSgYwfrFh2U1IW',
      createdAt: '1694792186448',
      description: '',
      title: '',
      isCompleted: true,
      updatedAt: '1694792197288',
    },
    {
      id: 'nL69dRR9S1kcKJJoYhUU',
      createdAt: '1694710896986',
      description:
        "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
      title: 'What is Lorem Ipsum?',
      isCompleted: true,
      updatedAt: '1694731701362',
    },
  ];
  let tasks$: Observable<Task[]>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [TaskService<Task>],
    });
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(ListComponent);
    component = fixture.componentInstance;
    tasks$ = of(tasksMock);

    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should match snapshot', () => {
    const compiled = fixture.nativeElement as HTMLElement;
    expect(compiled).toMatchSnapshot();
  });

  test('should get task list', () => {
    tasks$.subscribe((data) => {
      expect(data).toBe(tasksMock);
    });

    tasks$
      .pipe(
        map((tasks) => tasks.filter((task: Task) => task.isCompleted)),
        tap((tasks: Task[]) =>
          tasks.forEach((task: Task) => {
            expect(task.isCompleted).toBeTruthy();
          })
        )
      )
      .subscribe();
  });

  test('should get a task list completed', () => {
    tasks$
      .pipe(
        map((tasks: Task[]) => tasks.filter((task: Task) => task.isCompleted)),
        tap((tasks: Task[]) =>
          tasks.forEach((task: Task) => expect(task.isCompleted).toBeTruthy())
        )
      )
      .subscribe();
  });
});
