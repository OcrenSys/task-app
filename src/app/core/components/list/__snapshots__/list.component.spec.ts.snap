// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <mat-card
    class="mat-mdc-card mdc-card card mt-2 mb-2"
  >
    <mat-card-header
      class="mat-mdc-card-header card-header"
    >
      <div
        class="mat-mdc-card-header-text"
      />
      <mat-card-title-group
        class="mat-mdc-card-title-group"
      >
        <div>
          <mat-card-title
            class="mat-mdc-card-title"
          >
            Completadas
          </mat-card-title>
          <mat-card-subtitle
            class="mat-mdc-card-subtitle"
          >
            Lista de tareas completadas Â  0/0
          </mat-card-subtitle>
        </div>
      </mat-card-title-group>
    </mat-card-header>
    <mat-card-content
      class="mat-mdc-card-content"
    >
      <div
        class="table-content"
      >
        <mat-list
          aria-disabled="false"
          class="mat-mdc-list mat-mdc-list-base mdc-list"
          role="list"
        >
          
        </mat-list>
      </div>
    </mat-card-content>
  </mat-card>
</div>
`;
